### PR TITLE
Added skip upgrade flag

### PR DESCRIPTION
**Pull request by sahith-narahari**
#5 
- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/develop/CONTRIBUTING.md#pr-targeting))

- [X] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [x] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [ ] Added entries in `PENDING.md` with issue # 
- [ ] rereviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: x/upgrade/abci.go [hunk 1]
```diff
@@ -3,22 +3,34 @@ package upgrade
...
+// The purpose is to ensure the binary is switched EXACTLY at the desired block, and to allow
 // a migration to be executed if needed upon this switch (migration defined in the new binary)
 func BeginBlocker(k Keeper, ctx sdk.Context, _ abci.RequestBeginBlock) {
+
 	plan, found := k.GetUpgradePlan(ctx)
 	if !found {
 		return
 	}
+
+	skipUpgrade := viper.Get(FlagUnsafeSkipUpgrade).(bool)
```
Comment (ethanfrey) [inline x/upgrade/abci.go]: I don't like reading viper here, makes this core logic hard to test.
If reading viper could be moved into the Module, and just pass it as an arg to this function, I would be happier.

File: x/upgrade/abci.go [hunk 2]
```diff
@@ -3,22 +3,34 @@ package upgrade
...
+	if skipUpgrade {
```
Comment (ethanfrey) [inline x/upgrade/abci.go]: This logic looks proper. However, I think we need one more condition. All nodes must execute this `ClearUpgradePlan` at the same block.

The solution I can think of is to place this `if skipUpgrade` block inside `if plan.ShouldExecute()` - at the top. Thus, when it comes time to execute, it will clear the plan instead and continue, using the came coordination logic we use for applying the upgrade.

File: x/upgrade/abci.go (hunk 1 above)
Comment (sahith-narahari) [reply x/upgrade/abci.go]: I feel it's hard to read this flag elsewhere,as we're writing into BeginBlocker which makes it hard to pass the flag as arguments.
Any suggesstions here.?

File: x/upgrade/abci.go (hunk 2 above)
Comment (sahith-narahari) [reply x/upgrade/abci.go]: Yes,changed accordingly.

File: x/upgrade/abci.go (hunk 1 above)
Comment (ethanfrey) [reply x/upgrade/abci.go]: I would read it here: https://github.com/regen-network/cosmos-sdk/blob/regen-network/upgrade-module/x/upgrade/module.go#L127-L129

This is an untested shim, and can pass more arguments in. This actual `BeginBlocker` function has lots of business logic to test and should be unit-testable.

(I don't change the architecture at all, just move the call one level up)

**Comment by ethanfrey** [comment]
I also finished off the upgrade modules PR and submitted it for final review (and hopefully merge). Can you rebase on that branch once you complete the above changes? Or if you feel uncomfortable with rebase, ping me when this is ready and I will do so.

https://github.com/regen-network/cosmos-sdk/commits/regen-network/upgrade-module

File: x/upgrade/abci.go (hunk 1 above)
Comment (sahith-narahari) [reply x/upgrade/abci.go]: Got it,I felt it's bad to refactor BeginBlocker format against other modules.But if module.go/BeginBlock is an untested shim,it should totally be fine.
Will add test scripts accordingly to this.

File: x/upgrade/abci_test.go [hunk 3]
```diff
@@ -23,6 +25,7 @@ type TestSuite struct {
 	module  module.AppModule
 	ctx     sdk.Context
 	cms     store.CommitMultiStore
+	FlagUnsafeSkipUpgrade string
```
Comment (ethanfrey) [inline x/upgrade/abci_test.go]: Why not use a boolean here instead?

File: x/upgrade/abci_test.go [hunk 4]
```diff
@@ -38,6 +41,7 @@ func (s *TestSuite) SetupTest() {
 	s.cms.MountStoreWithDB(key, sdk.StoreTypeIAVL, db)
 	_ = s.cms.LoadLatestVersion()
 	s.ctx = sdk.NewContext(s.cms, abci.Header{Height: 10, Time: time.Now()}, false, log.NewNopLogger())
+	s.FlagUnsafeSkipUpgrade = FlagUnsafeSkipUpgrade
```
Comment (ethanfrey) [inline x/upgrade/abci_test.go]: And then do `viper.Set(FlagUnsafeSkipUpgrade, s.skipUpgrade)` here, rather than inside the tests code.

Otherwise, you can remove the calls to `module.BeginBlock` and just call the BeginBlocker directly and avoid all use of viper, which I think is cleaner. I actually only use the module here to try to increase the test coverage a bit more. Not sure if it makes sense to keep using it for BeginBlock or just call the function directly.

Comment (ethanfrey) [reply x/upgrade/abci_test.go]: In any case, this should be set for all calls in the Setup, not in individual tests, where it is easy to forget

File: x/upgrade/abci_test.go [hunk 5]
```diff
@@ -185,6 +191,35 @@ func (s *TestSuite) TestPlanStringer() {
...
+	req := abci.RequestBeginBlock{Header: newCtx.BlockHeader()}
+	err := s.handler(s.ctx, SoftwareUpgradeProposal{Title: "prop", Plan: Plan{Name: "test", Height: s.ctx.BlockHeight() + 1}})
+	s.Require().Nil(err)
+
+	s.T().Log("Verify if skip upgrade flag clears upgrade plan")
+	viper.Set(s.FlagUnsafeSkipUpgrade, true )
+	s.Require().NotPanics(func() {
+		s.module.BeginBlock(newCtx, req)
+	})
+	s.VerifyCleared(s.ctx)
```
Comment (ethanfrey) [inline x/upgrade/abci_test.go]: Please also add a check that this was not actually executed - that the names plan is not listed in done either

File: x/upgrade/abci_test.go [hunk 6]
```diff
@@ -185,6 +191,35 @@ func (s *TestSuite) TestPlanStringer() {
...
 
+func (s *TestSuite) TestSkipUpgrade()  {
+	newCtx := sdk.NewContext(s.cms, abci.Header{Height: s.ctx.BlockHeight() + 1, Time: time.Now()}, false, log.NewNopLogger())
+	req := abci.RequestBeginBlock{Header: newCtx.BlockHeader()}
+	err := s.handler(s.ctx, SoftwareUpgradeProposal{Title: "prop", Plan: Plan{Name: "test", Height: s.ctx.BlockHeight() + 1}})
+	s.Require().Nil(err)
+
+	s.T().Log("Verify if skip upgrade flag clears upgrade plan")
+	viper.Set(s.FlagUnsafeSkipUpgrade, true )
+	s.Require().NotPanics(func() {
```
Comment (ethanfrey) [inline x/upgrade/abci_test.go]: If we VerifyCleared after this call, we should VerifySet (or whatever the name is) before the call, to make clear that we are testing this functionality

File: x/upgrade/abci_test.go (hunk 3 above)
Comment (sahith-narahari) [reply x/upgrade/abci_test.go]: We're not declaring the bool here,instead declaring the key and assigning a value. Coz we're reading the value of flag in BeginBlock,so we need to just assign the key and check it's being read in beginblock

File: x/upgrade/abci_test.go (hunk 4 above)
Comment (sahith-narahari) [reply x/upgrade/abci_test.go]: This(viper.set) should make it cleaner,will change.
I think we need to check the flag is being read in BeginBlock rather than just testing BeginBlocker

File: x/upgrade/abci_test.go (hunk 5 above)
Comment (sahith-narahari) [reply x/upgrade/abci_test.go]: Yeah,missed that.Will change.

File: x/upgrade/abci_test.go (hunk 4 above)
Comment (ethanfrey) [reply x/upgrade/abci_test.go]: Fair enough

Comment (ethanfrey) [reply x/upgrade/abci_test.go]: Okay

File: x/upgrade/abci_test.go (hunk 6 above)
Comment (sahith-narahari) [reply x/upgrade/abci_test.go]: Changed this,added verify set

File: x/upgrade/abci.go [hunk 7]
```diff
@@ -9,17 +9,29 @@ import (
...
 	if plan.ShouldExecute(ctx) {
+		//To make sure clear upgrade is executed at the same block
+		if skipUpgrade {
+			// If skip upgrade has been set, we clear the upgrade plan
+			skipUpgradeMsg := fmt.Sprintf("UPGRADE \"%s\" SKIPPED at %s: %s", plan.Name, plan.DueAt(), plan.Info)
+			ctx.Logger().Info(skipUpgradeMsg)
+			k.ClearUpgradePlan(ctx)
+			return
+		}
+
```
[the remark this answers is not in the document]
Comment (anilCSE) [reply x/upgrade/abci.go]: @aaronc yes that is true. Instead of taking height and cross checking, I propose to clear the flag (unset) once it is used. Because if we are starting the node with skip-upgrade flag, we intend to skip that particular upgrade and the flag should not carry any value after that process.

[the remark this answers is not in the document]
Comment (ethanfrey) [reply x/upgrade/abci.go]: Anil answer seems reasonable, assuming people manually set this flag. If they start putting it in the systems restart script, then it is problematic. But that is a bigger problem.

I see Aaron's approach as safer, but also harder to use. 

We should consider this will be used in emergency situations and there will be a long period before a valid upgrade proposal. I am not sure which is better, but having user stories in mind can help design.

As an alternative, I could suggest we include the name of the plan instead of height but use it the same way

As another alternative, it could take an optional height field, which will enforce Aaron's logic present, and default to Anils proposal (perform once and unset flag) if left as default (0). Here I would use 0 as skip any plan, positive as only skip at height x, and negative as flag used, ignore it.

[the remark this answers is not in the document]
Comment (aaronc) [reply x/upgrade/abci.go]: Here are some user stories:

* the most common - an upgrade is supposed to happen, but an error occurs during the upgrade handler and thus can't proceed, the validators agree to use `--unsafe-abort-upgrade` to keep going - they do know the height at which this happens
* validators agree in advance to skip an upgrade outside of governance - they may not know the height, BUT they can still want until the upgrade time hits and disable automatic upgrades, and then they can use the height to continue
* someone is syncing a full node from genesis - they need to know which upgrades to skip because these happen outside of the block history. In this case, they need to know the heights because the same named upgrade could have occurred later after a failure - i.e. say `upgrade1` was scheduled for height 5000 and failed, later `upgrade1` got rescheduled for 10000 and succeeds, so we only want `--unsafe-skip-upgrade 5000` not `--unsafe-skip-upgrade upgrade1` for full nodes so they can sync from genesis

So I know it is a bit more complex, but from this the solution that covers the most use cases and also seems the safest/most correct is we can allow multiple `--unsafe-skip-upgrade <height>` flags to handle potentially multiple aborted upgrades in a chain history - in particular this helps with syncing a full node from genesis which is a fairly common use case.

Thoughts?

[the remark this answers is not in the document]
Comment (anilCSE) [reply x/upgrade/abci.go]: I agree with @aaronc, this works best considering all the user stories. Instead of handling each case differently, `--unsafe-skip-upgrade <height>` can work commonly to handle all the cases

[the remark this answers is not in the document]
Comment (anilCSE) [reply x/upgrade/abci.go]: > * validators agree in advance to skip an upgrade outside of governance - they may not know the height, BUT they can still want until the upgrade time hits and disable automatic upgrades, and then they can use the height to continue

Regarding this, is there a way to disable automatic upgrades now?

[the remark this answers is not in the document]
Comment (aaronc) [reply x/upgrade/abci.go]: Well automatic upgrades can be disabled by either 1) not using `cosmosd` at all or 2) not providing `cosmosd` with an upgrade binary and not setting `DAEMON_ALLOW_DOWNLOAD_BINARIES`. So if you're either not using cosmosd or cosmosd has nothing to upgrade to, the binary should just panic at that height and the operator needs to intervene.

File: server/start.go [hunk 8]
```diff
@@ -78,8 +78,8 @@ which accepts a path for the resulting pprof file.
 		FlagMinGasPrices, "",
 		"Minimum gas prices to accept for transactions; Any fee in a tx must meet this minimum (e.g. 0.01photino;0.0001stake)",
 	)
-	cmd.Flags().Bool(FlagSkipUpgrade, false, "Skip current software upgrade to continue the old binary")
-	cmd.Flags().Uint64(FlagHaltHeight, 0, "Block height at which to gracefully halt the chain and shutdown the node")
+	cmd.Flags().Uint64(FlagSkipUpgrade, 0, "Skip current software upgrade to continue the old binary")
+	cmd.Flags().Int64(FlagHaltHeight, -1, "Block height at which to gracefully halt the chain and shutdown the node")
```
Comment (ethanfrey) [inline server/start.go]: why is this changed? I don't see any corresponding change to the logic

Comment (sahith-narahari) [reply server/start.go]: It was a typo on my side,the latest commit should resolve this.
Flag skip upgrade was changed to Int64 as ctx.Blockheight() is of type int64

Comment (ethanfrey) [reply server/start.go]: Yes, I agree with this change, and the rest of the diff looks quite clean.

**Comment by ethanfrey** [comment]
It looks good. Just one comment based on @aaronc requirement:

> So I know it is a bit more complex, but from this the solution that covers the most use cases and also seems the safest/most correct is we can allow multiple --unsafe-skip-upgrade <height> flags to handle potentially multiple aborted upgrades in a chain history - in particular this helps with syncing a full node from genesis which is a fairly common use case.

This implementation only handles one copy of the flag. There is a function [viper.GetIntSlice](https://godoc.org/github.com/spf13/viper#Viper.GetIntSlice) which can be used here (unfortunately there is not an int64, you must cast). IIRC, it allows both `--skip-upgrade=123 --skip-upgrade=456` and `--skip-upgrade=123,456`.

**Comment by ethanfrey** [comment]
To update the flags, you can use `IntSlice` instead of `Uint64`: https://godoc.org/github.com/spf13/pflag#IntSlice

cobra does allow Uint64Slice and such, but as viper does not support that (and that is how we are reading the values), it seems better to stick to the best supported path.

File: server/start.go [hunk 9]
```diff
@@ -77,6 +78,7 @@ which accepts a path for the resulting pprof file.
 		FlagMinGasPrices, "",
 		"Minimum gas prices to accept for transactions; Any fee in a tx must meet this minimum (e.g. 0.01photino;0.0001stake)",
 	)
+	cmd.Flags().IntSlice(FlagUnsafeSkipUpgrade, []int{-1}, "Skip current software upgrade to continue the old binary")
```
Comment (ethanfrey) [inline server/start.go]: What about simply using `[]int{}` here? We don't need a flag value, and I would assume this default if no flags are passed.

File: x/upgrade/abci.go [hunk 10]
```diff
@@ -9,17 +9,29 @@ import (
...
 	plan, found := k.GetUpgradePlan(ctx)
 	if !found {
 		return
 	}
+
 	if plan.ShouldExecute(ctx) {
+		//To make sure clear upgrade is executed at the same block
+		if k.Contains(skipUpgradeHeightArray, ctx.BlockHeight()) {
+			// If skip upgrade has been set, we clear the upgrade plan
+			skipUpgradeMsg := fmt.Sprintf("UPGRADE \"%s\" SKIPPED at %s: %s", plan.Name, plan.DueAt(), plan.Info)
```
Comment (ethanfrey) [inline x/upgrade/abci.go]: I would actually use `Height` instead of `DueAt` here, I think `DueAt` may return a timestamp as well.
Normally, `DueAt()` is nice, but since this skip is triggered by a flag with an exact block height, we should show this, so it lines up with the flag.

Comment (ethanfrey) [reply x/upgrade/abci.go]: Also, I note that there is no way to securely skip an upgrade that is triggered by time. This would need a wide range of skip flags to cover many possible heights. Once it fails upgrading, however, we know which block it is in and can skip, so this still works as an "abort upgrade after fail" approach, just not a pre-emptative upgrade.

I think handling this is out of scope of the PR, I just want to mention it, maybe Aaron wants to open a new issue to address that. (Possibly deprecating use of block time trigger)

File: x/upgrade/abci_test.go [hunk 11]
```diff
@@ -38,6 +42,9 @@ func (s *TestSuite) SetupTest() {
 	s.cms.MountStoreWithDB(key, sdk.StoreTypeIAVL, db)
 	_ = s.cms.LoadLatestVersion()
 	s.ctx = sdk.NewContext(s.cms, abci.Header{Height: 10, Time: time.Now()}, false, log.NewNopLogger())
+	s.FlagUnsafeSkipUpgrade = FlagUnsafeSkipUpgrade
+	viper.Set(FlagUnsafeSkipUpgrade, []int{-1})
```
Comment (ethanfrey) [inline x/upgrade/abci_test.go]: If you update the flag default above, please adjust here as well.

File: x/upgrade/abci_test.go [hunk 12]
```diff
@@ -151,7 +174,8 @@ func (s *TestSuite) TestCanClear() {
 	err := s.handler(s.ctx, SoftwareUpgradeProposal{Title: "prop", Plan: Plan{Name: "test", Time: time.Now()}})
 	s.Require().Nil(err)
 
-	s.handler(s.ctx, CancelSoftwareUpgradeProposal{Title: "cancel"})
+	err = s.handler(s.ctx, CancelSoftwareUpgradeProposal{Title: "cancel"})
```
Comment (ethanfrey) [inline x/upgrade/abci_test.go]: :+1: good to check those returned errors

File: x/upgrade/abci_test.go [hunk 13]
```diff
@@ -185,6 +209,162 @@ func (s *TestSuite) TestPlanStringer() {
...
+
+	var i int64
+
+	s.T().Log("Verify if skip upgrade flag clears upgrade plan in both cases")
+	viper.Set(s.FlagUnsafeSkipUpgrade, []int{int(s.ctx.BlockHeight() + 1), int(s.ctx.BlockHeight() + 10)})
+	s.VerifySet()
+
+	s.VerifyConversion(viper.GetIntSlice(s.FlagUnsafeSkipUpgrade))
+
+	for i = 1; i <= s.ctx.BlockHeight(); i++ {
```
Comment (ethanfrey) [inline x/upgrade/abci_test.go]: It would be nice to make it more clear we tested the first upgrade here... I was confused why you stopped one block early, but then saw `newCtx.WithBlockHeight(s.ctx.BlockHeight() + i)`. which actually doesn't make too much sense... (doing from blockheight+1 to double blockheight).

I would:
* rename `i` to `h`
* make `h` exactly the block height you want `newCtx.WithBlockHeight(h)`
* adjust the for loop to whatever blocks you want to cover.

Not that this code is wrong persay, just not clear.
(Maybe you do want to test if the flag is set for a block that has no upgrade, that is a fine test, and you do it here, probably, it is just not clear)

File: x/upgrade/abci_test.go [hunk 14]
```diff
@@ -185,6 +209,162 @@ func (s *TestSuite) TestPlanStringer() {
...
+	}
+
+	s.T().Log("Verify a second proposal also is being cleared")
+	err = s.handler(s.ctx, SoftwareUpgradeProposal{Title: "prop2", Plan: Plan{Name: "test2", Height: s.ctx.BlockHeight() + 10}})
+	s.Require().Nil(err)
+	s.Require().NotPanics(func() {
+		s.module.BeginBlock(newCtx, req)
+	})
+
+	//To ensure verification is being done only after both uprades are cleared
```
Comment (ethanfrey) [inline x/upgrade/abci_test.go]: Good test to check it was all cleaned up

File: x/upgrade/abci_test.go [hunk 15]
```diff
@@ -185,6 +209,162 @@ func (s *TestSuite) TestPlanStringer() {
...
+
+	var i int64
+
+	s.T().Log("Verify if skip upgrade flag clears upgrade plan in one case and does upgrade on another")
+	viper.Set(s.FlagUnsafeSkipUpgrade, []int{int(s.ctx.BlockHeight() + 1)})
+	s.VerifySet()
+
+	s.VerifyConversion(viper.GetIntSlice(s.FlagUnsafeSkipUpgrade))
+
+	for i = 1; i < s.ctx.BlockHeight(); i++ {
```
Comment (ethanfrey) [inline x/upgrade/abci_test.go]: Same comment with for loop as above

File: x/upgrade/abci_test.go [hunk 16]
```diff
@@ -185,6 +209,162 @@ func (s *TestSuite) TestPlanStringer() {
...
+		newCtx = newCtx.WithBlockHeight(s.ctx.BlockHeight() + i)
+		s.Require().NotPanics(func() {
+			s.module.BeginBlock(newCtx, req)
+		})
+	}
+
+	s.T().Log("Verify the second proposal is not skipped")
+	err = s.handler(s.ctx, SoftwareUpgradeProposal{Title: "prop2", Plan: Plan{Name: "test2", Height: s.ctx.BlockHeight() + 10}})
+	s.Require().Nil(err)
+	newCtx = newCtx.WithBlockHeight(newCtx.BlockHeight() + 1)
```
Comment (ethanfrey) [inline x/upgrade/abci_test.go]: This is very confusing.`newCtx` keeps getting updated in the for loop, so no idea what the height is, relative to `s.ctx.BlockHeight()`. 

For these ones where you test a lot, you can simply define `startHeight` above, and make everything relative to that. (Or use `s.ctx.BlockHeight()` instead of `startHeight`). 

I mean, I would expect this test to pass when this line read `newCtx = s.ctx.WithBlockHeight(s.ctx.BlockHeight() + 1)`

File: x/upgrade/abci_test.go [hunk 17]
```diff
@@ -185,6 +209,162 @@ func (s *TestSuite) TestPlanStringer() {
...
+	s.VerifyConversion(viper.GetIntSlice(s.FlagUnsafeSkipUpgrade))
+
+	for i = 1; i < s.ctx.BlockHeight(); i++ {
+		newCtx = newCtx.WithBlockHeight(s.ctx.BlockHeight() + i)
+		s.Require().NotPanics(func() {
+			s.module.BeginBlock(newCtx, req)
+		})
+	}
+
+	//A new proposal with height in skip
```
Comment (ethanfrey) [inline x/upgrade/abci_test.go]: Same as above, both for lop, and `newCtx = newCtx.WithBlockHeight(newCtx.BlockHeight() + 1) ` below

File: x/upgrade/internal/keeper/utils.go [hunk 18]
```diff
@@ -0,0 +1,18 @@
+package keeper
+
+func (k Keeper) ConvertIntArrayToInt64(blockHeight []int) []int64 {
```
Comment (ethanfrey) [inline x/upgrade/internal/keeper/utils.go]: Looks fine

File: server/start.go (hunk 9 above)
Comment (sahith-narahari) [reply server/start.go]: That should be better, will use empty array as default

File: x/upgrade/abci.go (hunk 10 above)
Comment (sahith-narahari) [reply x/upgrade/abci.go]: Yes, even I had this concern.Block height is a better option for any upgrade and all pervious upgrades mentioned a height on which the state will be dumped.
@aaronc

File: x/upgrade/abci_test.go (hunk 13 above)
Comment (sahith-narahari) [reply x/upgrade/abci_test.go]: Over second thought,I feel this looping can be avoided and BeginBlock can be run only for the heights mentioned in the proposals. Do you think I'll miss some case if I do that, I don't see any issues

Comment (ethanfrey) [reply x/upgrade/abci_test.go]: It seems fine.
I would make sure to run every case in the following matrix at least once:

* upgrade=no, skip=no
* upgrade=yes, skip=no
* upgrade=no, skip=yes
* upgrade=yes, skip=yes

Beyond that, no need to check any blocks (and this matrix doesn't have to be run in every text case, but each at least once in the suite).

**Comment by sahith-narahari** [comment]
Closing this as it's already merged upstream
